### ISSUE TITLE
Tracing the boundary curve of an image region

### ISSUE THREAD

**Issue by dcbb** [scikit-image/scikit-image#1131]
I need to trace the boundary curve of an image region enumerated by `measure.regionprops`, similar to [`bwboundaries`](http://www.mathworks.de/de/help/images/ref/bwboundaries.html) in Matlab.

By boundary curve, I mean a list of border pixels of the region in, say, clockwise direction around the region's perimeter, such that I can, for example, represent the region with a polygon. I need the exact coordinates of all border pixels, not a convex hull approximation. I am currently using `measure.find_contours`, but this is an algorithmic overkill for simple binary boundary tracing (and slow).

Issue created as suggested by Stefan van der Walt on [StackOverflow](http://stackoverflow.com/questions/25379752/how-can-i-extract-the-boundary-curve-of-an-image-region-in-scikit-image#comment39900895_25379752).

**Comment by k88kramer** [comment]
This is a question repeatedly asked on the internet.  There are methods to find these region borders in MATLAB, openCV and mahotas that I have found.  Is there a reason why a "border coordinates" function for labeled regions is not included in regionprops in scikit-image?

**Comment by jni** [comment]
Hi @dcbb and @k88kramer! We haven't forgotten about you! 

@k88kramer the usual reason for something not being in skimage is lack of time. Please remember that this is a community-driven project and unless someone who needs something implements it, it will just sit on the backburner.

I'm looking for a minor hacking project so if you can point me to a description of the algorithm, I might tackle this one soon...

**Comment by k88kramer** [comment]
Thanks for your reply.

In scimage.measure.regionprops there are many properties that can be 
accessed such as "coords" which is an array of all the pixel coordinates 
within a labeled region, and "perimeter" which is the computed length of 
the region perimeter.  However there is no property for the perimeter 
pixel coordinates.

It seems to me that the actual pixel coordinates of the region perimeter 
would be needed internally in skimage.measure.regionprops to provide 
some other properties such as those listed above.  Since the data is 
already there internally it would be useful if it were made available as 
a property.

Something like:

perimeter_coords: (N,2) ndarray

Ordered array of the pixel coordinates (x, y) along the perimeter of the 
region.

I am only learning python and am not familiar with the internals of 
skimage.  If you do come up with an implementation of exposing 
"perimeter_coords" I would be very interested in seeing your code. It 
would help me learn so that maybe someday I could also contribute.

Thanks again for your reply and for all the work on a great project like 
scikits-image.

Jack Kramer

On 11/18/2015 01:36 AM, Juan Nunez-Iglesias wrote:
>
> Hi @dcbb <https://github.com/dcbb> and @k88kramer 
> <https://github.com/k88kramer>! We haven't forgotten about you!
>
> @k88kramer <https://github.com/k88kramer> the usual reason for 
> something not being in skimage is lack of time. Please remember that 
> this is a community-driven project and unless someone who needs 
> something implements it, it will just sit on the backburner.
>
> I'm looking for a minor hacking project so if you can point me to a 
> description of the algorithm, I might tackle this one soon...
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/scikit-image/scikit-image/issues/1131#issuecomment-157619915>.
>

-- 
Jack Kramer
Gainesville, Florida

**Comment by jni** [comment]
Hey @k88kramer!

> It seems to me that the actual pixel coordinates of the region perimeter
> would be needed internally in skimage.measure.regionprops

You would think so, but you would be wrong! =P That's actually where I first went to look. You can see the code [here](https://github.com/scikit-image/scikit-image/blob/master/skimage/measure/_regionprops.py#L513). To save you a click, the issue is that to compute the length of the perimeter, you need the coordinates, but they don't need to be sorted. To draw them, however, you need to sort them. And that's where I don't quite know how to proceed...

**Comment by k88kramer** [comment]
They for looking.  I used the contours of the regions as an 
approximation.  When I have time I will write a routine to trace the 
outline in pixels either from the region contours approximation or the 
regionprops coords property.

On 11/18/2015 10:57 PM, Juan Nunez-Iglesias wrote:
>
> Hey @k88kramer <https://github.com/k88kramer>!
>
>     It seems to me that the actual pixel coordinates of the region
>     perimeter
>     would be needed internally in skimage.measure.regionprops
>
> You would think so, but you would be wrong! =P That's actually where I 
> first went to look. You can see the code here 
> <https://github.com/scikit-image/scikit-image/blob/master/skimage/measure/_regionprops.py#L513>. 
> To save you a click, the issue is that to compute the length of the 
> perimeter, you need the coordinates, but they don't need to be sorted. 
> To draw them, however, you need to sort them. And that's where I don't 
> quite know how to proceed...
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/scikit-image/scikit-image/issues/1131#issuecomment-157943978>.
>

-- 
Jack Kramer
Gainesville, Florida

**Comment by sciunto** [comment]
@dcbb  @k88kramer Is it something like this that you are looking for?
![boundary](https://cloud.githubusercontent.com/assets/335370/17446685/a99a8364-5b4a-11e6-9db9-9fccf3ee2f39.png)


I'll submit a PR for that soon. Just need to clean up the code and write tests.

**Comment by stefanv** [comment]
See https://en.wikipedia.org/wiki/Boundary_tracing and
http://engr.case.edu/merat_francis/eecs490f07/lectures/lecture23.pdf

**Comment by jni** [comment]
@sciunto 👏 👏 👏

**Comment by giswqs** [comment]
I use the following code as a workaround to get boundary pixels

`import matplotlib.pyplot as plt
from scipy import ndimage
import numpy as np

a = np.zeros((50,50))
a[10:30,10:30] = 1
a[35:45,35:45] = 2

distance = ndimage.distance_transform_edt(a)

distance[distance != 1] = 0
plt.imshow(distance)
plt.show()

np.where(distance == 1)`

**Comment by soupault** [comment]
@sciunto would you like to share your code? I could improve and finalize it if needed.

**Comment by stefanv** [comment]
ping @sciunto ; we just ran into this issue again.  Do you have any code that we can work from?

**Comment by sciunto** [comment]
nope, sorry :(

**Comment by d-diaz** [comment]
@stefanv, you might be able to use the [`rasterio.features` module](https://rasterio.readthedocs.io/en/latest/api/rasterio.features.html). It's designed for geospatially-referenced images, which may not be included with your data, but you could probably inspect the source code and generalize if necessary. In particular, I'd suggest looking at [`rasterio.features.shapes`](https://github.com/mapbox/rasterio/blob/925718ea5cc61eaca46dcdb03739ecdde18f8e21/rasterio/features.py#L78). This returns polygons based on contiguous areas in an image.

**Comment by emmanuelle** [comment]
The big difficulty here is really the sorting of pixels, because just getting the boundary pixels can just be doing a logical xor (or a subtraction) between an object and its erosion.

**Comment by jklausing** [comment]
From the docs on Matlab's bwboundaries:

> The bwboundaries function implements the Moore-Neighbor tracing algorithm modified by Jacob's stopping criteria. This function is based on the boundaries function presented in the first edition of Digital Image Processing Using MATLAB, by Gonzalez, R. C., R. E. Woods, and S. L. Eddins, New Jersey, Pearson Prentice Hall, 2004.

For reference:
[Moore-Neighbor Tracing](http://www.imageprocessingplace.com/downloads_V3/root_downloads/tutorials/contour_tracing_Abeer_George_Ghuneim/moore.html)

**Comment by ElmWer** [comment]
A good appoximation to the boundary curve of an image region is the concave hull (see for example [Moreira & Santos](https://pdfs.semanticscholar.org/2397/17005c3ebd5d6a42fc833daf97a0edee1ce4.pdf) or [Figueira](https://towardsdatascience.com/the-concave-hull-c649795c0f0f).

A simple script using this approach can be found [here](https://drive.google.com/file/d/12Sk471gkHYzD0kKoThZYmA9dkz5WIWYQ/view?usp=sharing).

**Comment by stefanv** [comment]
Please note again the distinction between curve tracing and calculating all pixels belonging to the edge. The last problem is much easier, but not what we're looking for here.

**Comment by ElmWer** [comment]
Please note, the algorithm of Moreira&Santos returns the **boundary curve** (all pixels of the boundary of an image region in the right order) and not only the pixels belonging to the edge.

**Comment by lagru** [comment]
@stefanv suggested during the SciPy 2020 sprint that [this implementation of moore's algorithm](https://github.com/machine-shop/deepwings/blob/6526066a08843e1dc4c16063bf820e3975d8171a/deepwings/method_features_extraction/image_processing.py#L156-L245) might be a good starting or inspiration point. I agree, Cython would be useful to speed up the loops here.

**Comment by FirefoxMetzger** [comment]
I felt like hacking a bit tonight, so I threw together a poor man's boundary tracing algorithm. It finds all boundaries in the image and returns them in a list. Each boundary is returned as a list of arrays (like np.nonzero). Pixels are oriented in negative order: clockwise for outward-facing boundaries, anti-clockwise for inward-facing boundaries. For a donut, the outside would have clock-wise orientation, and the hole would have anti-clockwise orientation.

It's likely not as sophisticated as Moore's Boundary Tracing albeit being similar. A limitation is that the current implementation can't be parallelized trivially in a lock-free manner. It does find all boundaries in the image, so it might be a nice point for discussion.

Notebook:
https://github.com/FirefoxMetzger/ipynb_boundary_tracing/blob/master/trace_boundaries.ipynb

Example Image:
![image](https://user-images.githubusercontent.com/4402489/91666394-9e7ef500-eafc-11ea-9875-39ac53b56214.png)

**Comment by stefanv** [comment]
That's awesome @FirefoxMetzger!  Did you see the Moore code I linked to above?  It could be fun to compare the two implementations.

API wise, it would be nice to have a RegionProp object go in and get a boundary out.

**Comment by FirefoxMetzger** [comment]
@stefanv On what criteria would you compare the methods?

Regarding RegionProp, if I understand the current API correctly, you would simply pass `regionprop.image` into the function.

On a tangent: Do you have an actual use-case for such algorithms? I tried to come up with an example for myself but always thought of an alternative method that would achieve the same result with existing methods.

**Comment by stefanv** [comment]
Boundary smoothing via the FFT requires an ordered set of boundary coordinates.

You ask hard questions: no idea how to compare, but perhaps may spot invalid behavior?

**Comment by FirefoxMetzger** [comment]
Hm, not sure what constitutes invalid behavior; perhaps this thought is useful:

The main difference between my method and Moore's (when using "Jacob's stopping criteria" - what MATLAB does) is how we conceptualize (and thus track) the orientation of a pixel. Moor's method walks along an unoriented border (2D) and tracks the preceding pixel for orientation, I walk along an oriented border (3D) and thus don't need to track orientation.

The manifestation of this is that I will select the same position `(rr, cc)` multiple times if the orientation changes whereas Moore's method won't. For example

```
[ 0, 0, 0],
[ 0, 1, 0],
[ 0, 0, 0]
```

will produce `[(1, 1)]` whereas my method will produce `[(1, 1), (1, 1), (1, 1), (1, 1)]`. Which may seem weird at first glance, but makes more sense to me if we add orientation: Moore's: `[(1, 1, N)]` Mine: `[(1, 1, N), (1, 1, E), (1, 1, S), (1, 1, W)]`.

Moore's method selects pixels multiple times, too, however never consecutively. Consider

```
[0, 0, 0, 0, 0],
[0, 1, 0, 0, 0],
[0, 0, 2, 0, 0],
[0, 0, 0, 3, 0],
[0, 0, 0, 0, 0]
```

where I numbered the individual pixels in the region for convenience. Moore's method starting at `1` (North) will produce something like `[1, 2, 3, 2]`, choosing pixel `2` twice but pixel `1` and `3` only once. My method will select pixels for each side that faces a `0` and produce  `[1, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 1]` which again becomes clear once you add orientation  `[(1,N), (1,E), (2, N), (2,E), (3,N), (3,E), (3,S), (3,W), (2,S), (2,W), (1,S), (1,W)]`.

No idea which version is more desirable. I'd obviously choose mine, but that is likely a bias because it was my idea.

**Comment by crisluengo** [comment]
I came here through an old, recently bumped question on Stack Overflow.

This is my best explanation, with simple MATLAB code, for the standard algorithm to trace an object boundary:
https://www.crisluengo.net/archives/324/

Please don’t use an ad-hoc invention, it’s likely going to have an issue that doesn’t show up until someone tries to apply it in some degenerate case.

Listing a boundary pixel multiple times where the object is one pixel thick is important, if you need the boundary traced pixel by pixel. If you skip those steps, measurements computed from the polygon (or from the chain code) will be wrong.

Note that this representation of the object allows for a more precise estimate of the perimeter than having the unsorted list of boundary pixels:
https://www.crisluengo.net/archives/310/

**Comment by stefanv** [comment]
We were looking at established boundary tracing methods, as mentioned in https://github.com/scikit-image/scikit-image/issues/1131#issuecomment-237930802. I think it was the Moore tracing algorithm.

**Comment by stefanv** [comment]
Yes, we don't yet have the Moore algorithm included. The current mechanisms are stopgaps for visualization, but it's been on our todo list for a long time.

It'd be great if someone could spend some time on doing it properly.
